File: leetcodes/11. Container With Most Water/maxArea.java
class Solution {
  public int maxArea(int[] height) {
    // constraint: len >=2
    // input: array, output: integer
    // two pointers while loop, time O(n), space O(1)
    int n = height.length;
    int s = 0;
    int e = n - 1;
    int max_area = 0;
    while (s < e) {
      int w = e - s;
      int h = (height[s] > height[e]) ? height[e] : height[s];
      max_area = Math.max(max_area, w * h);
      if (height[s] <= height[e])
        s++;
      else
        e--;
    }
    return max_area;
  }
}

File: leetcodes/11. Container With Most Water/maxArea.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        h = len(height) - 1
        maxA = 0
        while(l < h):
          W = h - l
          maxH = min(height[l], height[h])
          maxA = max(maxA, maxH*W)
          if height[l] > height[h]: h-=1
          else: l += 1
        return maxA